File: src/com/kingmang/lazurite/parser/AST/Statements/ThrowStatement.java
package com.kingmang.lazurite.parser.AST.Statements;

import com.kingmang.lazurite.exceptions.OperationIsNotSupportedException;
import com.kingmang.lazurite.parser.AST.Expressions.Expression;
import com.kingmang.lazurite.patterns.visitor.ResultVisitor;
import com.kingmang.lazurite.patterns.visitor.Visitor;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.Setter;

@AllArgsConstructor
public class ThrowStatement implements Statement {
    @Getter
    @Setter
    private String type;
    @Getter
    @Setter
    private Expression expr;


    @Override
    public void execute() {
        throw new OperationIsNotSupportedException(type, expr.eval().toString());
    }

    @Override
    public String toString() {
        return "throw " + type + " " + expr;
    }


    @Override
    public void accept(Visitor visitor) {}

    @Override
    public <R, T> R accept(ResultVisitor<R, T> visitor, T input) {
        return null;
    }
}

File: src/main/java/com/kingmang/lazurite/parser/ast/statements/ThrowStatement.kt
package com.kingmang.lazurite.parser.ast.statements

import com.kingmang.lazurite.core.CallStack
import com.kingmang.lazurite.exceptions.FileInfo
import com.kingmang.lazurite.exceptions.LzrTracedException
import com.kingmang.lazurite.exceptions.LzrTracedException.TraceInfo
import com.kingmang.lazurite.parser.ast.expressions.Expression
import com.kingmang.lazurite.patterns.visitor.ResultVisitor
import com.kingmang.lazurite.patterns.visitor.Visitor

class ThrowStatement(
    private val type: String,
    private val expr: Expression,
    private val file: FileInfo?
) : Statement {
    override fun execute() =
        throw LzrTracedException(this.type, this.expr.eval().toString(), TraceInfo(this.file, CallStack.getCalls()))

    override fun toString(): String =
        "throw ${this.type} ${this.expr}"

    override fun accept(visitor: Visitor) =
        Unit

    override fun <R, T> accept(visitor: ResultVisitor<R, T>, input: T): R? =
        null
}